### DISCUSSION TITLE
Separate Next.js from Express.js in my full-stack app

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#10658] [Help]
Hello,

I have a question. Is it possible to separate (isolate) Next.js (let's say in `/src/client`) from Express.js (let's say in `/src/server`, so that there will be 2 `package.json` and 2 `node_modules`, and the client (Next.js) can be started independently in a different port than the server (Express.js)? I know Next.js is Server-Side Rendering, but it is still client stuff (understanding "client" here as front-end development). I have already done all the isolating stuff, I got it working, but the problem I have now is how to allow Express.js to manage the routes (not defined before in the Express.js `server.js` file) by Next.js.

So when writing a single `server.js` file which initiates Express.js with Next.js we include this line:
`app.all('*', (req, res) => requestHandler(req, res))`;
Here `requestHandler` is a handler defined by Next.js, but as I am not including or initiating Next.js in this `server.js` file I would need something just like it is done with create-react-app and Express.js (which can be easily isolated):
`app.get('*', (req, res) => res.sendFile(path.resolve(__dirname, 'client', 'build', 'index.html')));`

Regards,
Néstor Llamas

**Comment by game7** [comment]
assuming that you would like to maintain a single repo, you can probably isolate the two using subdirectories such as `/client` and `/server`, each with their own `package.json` and `/node_modules`.  This would allow sharing of common code between the two while keeping the dependencies and configuration independent of one another.

**Comment by ghost** [comment]
Yes, I've already done that. But you didn't answer to my question. Did you read it properly?

>  I have already done all the isolating stuff, I got it working, but the problem I have now is how to allow Express.js to manage the routes (not defined before in the Express.js server.js file) by Next.js.

**Comment by game7** [comment]
sorry -- looks like didn't read it thoroughly and answered way to quickly 🙄

It's been a while since I've used custom server with Next -- we switched to API routes when they were added in v9.  I might be able to help here but am not sure I understand the overall goal.  As you have isolated the client and server it sounds like you would also like to run them in isolation, in which case Express would not need to be aware of the Next routes and could act as an API server for your client -- just startup express.

Curious -- what is the motivation for isolating the client and the server?  Reason I ask is because I currently work on a few different "full stack" Next applications and haven't had any issues caused by mixing the client-specific and server-specific dependencies within a single package.json delivered by a single run-time -- that is, as long as you don't try to import something node-specific into a browser context and vice-versa.  

Each of these application's use a few express-compatible plugins (such as cookie-session) that we were able to attach to Next's req/res pipeline so that we could avoid the need for a custom server.  Life has been much easier since we moved away from custom servers.

Interestingly, Next 9.2.3 (canary) actually supports patterns that allow server-specific code to be written directly within page component files (`export async function unstable_getServerProps()`) where it is removed and called via fetch when running within the browser.

I know this response does not directly answer your question but figure it wouldn't hurt to share our progression from custom server to API routes

**Comment by ghost** [comment]
Yes, I was considering migrating my current web app from Next.js + Express.js to just Next.js (+ API routes). But in case I continue as I am, maybe because Express.js is still a more powerful server, is there any solution to my question?

**Reply by sarapowers** [reply]
Hey! I was browsing this next.js discussion board and your question caught my eye. Me and a few other open source developers recently created [Connext-js](https://github.com/oslabs-beta/connext-js) because we missed Express style middleware in Next.js apps but didn't want the added bulk of a custom server. It might help you!

**Comment by michaelparkadze** [comment]
Hey, did you find a solution for this situation? I am also looking to have one repo for next.js client side and another for the server side but after hours of looking for a solution still can not find a solid way to implement this.

**Comment by pokhiii** [comment]
+1
